### PR TITLE
[v6r16] REBASE new multi protocol merge

### PR DESCRIPTION
**Pull request by chaen**
Replaces https://github.com/DIRACGrid/DIRAC/pull/2668

This PR contains the next steps in order to implement multi protocol support for storage elements. It is however far from final now. Some cleanup are still needed, and will be done once the gfal2 plugins are production ready.

Basically, when performing an action on the StorageElement, instead of looping over all the protocol plugins, we loop over a filtered list. This list is built taking into account which action is taken (read vs write), and is also sorted according to lists defined in the CS. 
Two lists were already existing: RegistrationProtocols and ThirdPartyProtocols. They were moved into the Operation section where they belong to, and two more lists added: AccessProtocols and WriteProtocols. They all default to 'srm' and 'dips'. The utilities to fetch these lists are moved to DMSHelper, and each of them can be locally overwritten in the SE definition.

The negotiation for third party transfer is also improved: it takes into account all possible protocols the source SE is able to produce, and all protocols the target is able to receive as input.

Note that as of now, the behavior should not change, because the most used plugin (SRM2) still internally uses the DefaultProtocol list defined in the StorageElement section of the CS. This sort of bypasses all the logic. But as I said, once gfal2 is production ready, we can do the final cleanup.

This will have to be veeeeery carefully tested :-)

### CODE REVIEW

File: DataManagementSystem/Client/DataManager.py [hunk 1]
```diff
@@ -812,54 +811,74 @@ def __replicate( self, lfn, destSEName, sourceSEName = '', destPath = '', localC
...
+      if not replicationProtocols:
         possibleIntermediateSEs.append( candidateSE )
         log.debug( "No protocol suitable for replication found" )
         continue
 
-      log.debug( 'Found common protocols', replicationProtocol )
+      log.debug( 'Found common protocols', replicationProtocols )
 
       # THIS WOULD NOT WORK IF PROTO == file !!
+      # Why did I write that comment ?!
```
Comment (atsareg) [inline DataManagementSystem/Client/DataManager.py]: I guess this relates to the comment above, not below. Can be skipped

File: Resources/Storage/DIPStorage.py [hunk 2]
```diff
@@ -27,6 +27,10 @@
 
 class DIPStorage( StorageBase ):
 
+  _InputProtocols = ['file', 'dip', 'dips']
+  _OutputProtocols = ['dip', 'dips']
```
Comment (atsareg) [inline Resources/Storage/DIPStorage.py]: If these are immutable constant lists, then names must be in all capitals, e.g. INPUT_PROTOCOLS

File: DataManagementSystem/Client/DataManager.py (hunk 1 above)
Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: Yes it relates to what is before, but I am sure there was a reason, and I can't figure out why :-(

File: Resources/Storage/test/Test_FilePlugin.py [hunk 3]
```diff
@@ -153,7 +154,7 @@ def localPutFile( fn, size = 0 ):
     # wrong size
     res = localPutFile( self.existingFile, size = -1 )
     self.assert_( res['OK'], res )
-    self.assert_( self.existingFile in res['Value']['Failed'] )
+    self.assert_( self.existingFile in res['Value']['Failed'], res )
```
Comment (atsareg) [inline Resources/Storage/test/Test_FilePlugin.py]: assert_ form of unittest assertion is declared obsoleted in the unittest documentation. Use assertEqual or other assertXXX methods

**Comment by fstagni** [comment]
Just one thing, which has probably been discussed between you while I was on holiday: why v6r16? Especially considering the comment "this has to be very carefully tested".

**Comment by chaen** [comment]
What branch would you want to target ? This comment has been copied from the original PR almost a year ago, and passed on everytime I rebased. And it is true for any version it will go to: it has to be well tested.

**Comment by atsareg** [comment]
Well, as the review process of the v6r16 components is not very fast, this PR can also be included, provided that Christophe is available for the testing.

**Comment by fstagni** [comment]
There are again conflicts.

**Comment by chaen** [comment]
I will redo it. It starts to be slightly anoying :-)

**Pull request closed without merging**